packages/controllers/src/permissions/rpc-methods/requestPermissions.ts

```diff
@@ -0,0 +1,74 @@
+import { ethErrors } from 'eth-rpc-errors';
+import type { PermittedHandlerExport } from '@metamask/rpc-methods/types';
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+import { MethodNames } from '../utils';
+
+import { invalidParams } from '../errors';
+import type { PermissionConstraint, RequestedPermissions } from '../Permission';
+
+export const requestPermissionsHandler: PermittedHandlerExport<
+  RequestPermissionsHooks,
+  [RequestedPermissions],
+  PermissionConstraint[]
+> = {
+  methodNames: [MethodNames.requestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    requestPermissions: true,
+  },
+};
+
+type RequestPermissions = (
+  requestedPermissions: RequestedPermissions,
+  id: string,
+) => Promise<
+  [Record<string, PermissionConstraint>, { id: string; origin: string }]
+>;
+
+export type RequestPermissionsHooks = {
+  requestPermissions: RequestPermissions;
+};
+
+async function requestPermissionsImplementation(
+  req: JsonRpcRequest<[RequestedPermissions]>,
+  res: PendingJsonRpcResponse<PermissionConstraint[]>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { requestPermissions }: RequestPermissionsHooks,
```

Comment:
Could we rename this to `requestPermissionsForOrigin` so that it's easier to tell that it's bound by origin?

Reply:
68caeb2